### ISSUE TITLE
The result of demo.py is bad? Is something wrong?

### ISSUE THREAD

**Issue by giangnguyen2412** [jiasenlu/NeuralBabyTalk#43]
I tried run demo.py but the result is terrible, the textual caption looks good but visual caption is absolutely wrong. Did I miss sth here.

For example: `a [knapstack] riding a [aircraft] on a dirt road`

**Comment by jhhuang96** [comment]
@luulinh90s  Did you solve it about incorrect visual caption? I had the same weird result.

**Comment by giangnguyen2412** [comment]
I am sorry, may demo.py is wrong. I did not fix it. I contacted the author by email but he did not reply.

**Comment by jhhuang96** [comment]
```python
def decode_sequence_det(itow, itod, ltow, itoc, wtod, seq, bn_seq, fg_seq, vocab_size, opt):
    N, D = seq.size()
    det_idxs = []
    det_words = []
    out = []
    for i in range(N):
        txt = ''
        for j in range(D):
            if j >= 1:
                txt = txt + ' '
            ix = seq[i,j].item()
            if ix > vocab_size:
                det_word = itod[fg_seq[i,j].item()]
                det_class = itoc[wtod[det_word]]
                if opt.decode_noc and det_class in noc_object:
                    det_word = det_class
				
                if (bn_seq[i,j] == 1) and det_word in ltow:
                    word = ltow[det_word]
                else:
                    word = det_word
                word = '[ ' + word + ' ]'
                det_words.append(word)
                idx = ix - vocab_size - 1
                det_idxs.append(idx)
            else:
                if ix == 0:
                    break
                else:
                    word = itow[str(ix)]
            txt = txt + word
        out.append(txt)
    return out, det_idxs, det_words
```
I think maybe there is something wrong in this function "decode_sequence_det".

**Comment by giangnguyen2412** [comment]
I debugged a lot and a lot, but I concluded that we should ask the author. I can not fix it even with a week. I recommend you not to try. It will be time-consuming, and asking the author is the best way.

**Comment by giangnguyen2412** [comment]
I traced to the function you mentioned `decode_sequence_det` and trace the functions calling it but its confusing and hard to see. I think there is a mismatch between the proposals from Faster RCNN. May the given proposal file from the author is the error

**Comment by jhhuang96** [comment]
Which is the proposal file from the author? Did you mean "coco_detection.h5"?

**Comment by giangnguyen2412** [comment]
No. The proposals from RCNN. The file you mentioned is pretrained model

Vào 21:30, T.7, 8 Th6, 2019 jhhuang <notifications@github.com> đã viết:

**Comment by giangnguyen2412** [comment]
The hd5 file if I remember correctly

Vào 21:30, T.7, 8 Th6, 2019 jhhuang <notifications@github.com> đã viết:

**Comment by jhhuang96** [comment]
I run demo.py each time, the result of visual words are always different. It is weird!

**Comment by giangnguyen2412** [comment]
Can u show me the differrence between two runs?

**Comment by jhhuang96** [comment]
①tensor(391895): a [ vase ] riding a [ goat ] down a dirt road data/coco/images/val2014/COCO_val2014_000000391895.jpg
②tensor(391895): a [ air plane ] riding a [ pizza ] down a dirt road data/coco/images/val2014/COCO_val2014_000000391895.jpg
①tensor(60623): a [ cormorant ] sitting at a [ seaplane ] with a birthday [ automobile ] data/coco/images/val2014/COCO_val2014_000000060623.jpg
②tensor(60623): a [ worker ] sitting at a [ oxen ] with a birthday [ chair ] data/coco/images/val2014/COCO_val2014_000000060623.jpg

Demo.py only tests 2 images.

**Comment by giangnguyen2412** [comment]
You can run on your own images, let try it first

**Comment by jhhuang96** [comment]
a [ corgi ] standing in front of a projection screen

a [ politician ] standing in front of a projection screen

**Comment by YanLiqi** [comment]
> ```python
> def decode_sequence_det(itow, itod, ltow, itoc, wtod, seq, bn_seq, fg_seq, vocab_size, opt):
>     N, D = seq.size()
>     det_idxs = []
>     det_words = []
>     out = []
>     for i in range(N):
>         txt = ''
>         for j in range(D):
>             if j >= 1:
>                 txt = txt + ' '
>             ix = seq[i,j].item()
>             if ix > vocab_size:
>                 det_word = itod[fg_seq[i,j].item()]
>                 det_class = itoc[wtod[det_word]]
>                 if opt.decode_noc and det_class in noc_object:
>                     det_word = det_class
> 				
>                 if (bn_seq[i,j] == 1) and det_word in ltow:
>                     word = ltow[det_word]
>                 else:
>                     word = det_word
>                 word = '[ ' + word + ' ]'
>                 det_words.append(word)
>                 idx = ix - vocab_size - 1
>                 det_idxs.append(idx)
>             else:
>                 if ix == 0:
>                     break
>                 else:
>                     word = itow[str(ix)]
>             txt = txt + word
>         out.append(txt)
>     return out, det_idxs, det_words
> ```
> 
> I think maybe there is something wrong in this function "decode_sequence_det".

I think there is something wrong in this function too, have you solved this problem?

**Comment by Fanshia** [comment]
Did you figure it out?

**Comment by Mas-Y** [comment]
> You can run on your own images, let try it first

@giangnguyen2412 Did you try running on your own images? If yes, can you please share the steps

**Comment by Mas-Y** [comment]
> I run demo.py each time, the result of visual words are always different. It is weird!

@jhhuang96 @giangnguyen2412  Were you able to solve this issue. I am facing the same issue.

**Comment by giangnguyen2412** [comment]
> > You can run on your own images, let try it first
> 
> @giangnguyen2412 Did you try running on your own images? If yes, can you please share the steps

you can find how to run on your own image from [here](https://github.com/giangnguyen2412/Neural-Baby-Talk-python3)
